### ISSUE TITLE
Windows runners are consistently extremely slow compared to Linux and macOS

### ISSUE THREAD

**Issue by myitcv** [actions/runner-images#7320]
### Description

In the CUE project we are seeing incredibly slow run times on Windows 2022 runners. 

For a recent example see https://github.com/cue-lang/cue-trybot/actions/runs/4477594562/.

Roughly speaking, here are the numbers we are seeing averaged out across 20-30 builds per day.

| OS | `actions/checkout` | `go test` (cache hit) |
|-|-|-|
| `ubuntu-22.04` | 2-3 secs | 8 secs | 
| `macOS-11` | 5 secs | 9 secs | 
| `windows-2022` | 15-20 secs | 120+ secs | 

Notice that `actions/checkout` is consistently slower on Windows. Yes, there is a network access element to this, but it is minimal.

The `go test` comparison is the cleanest comparison. As indicated in the column heading, these are the timings for a full cache hit. i.e. no network access required, no rebuilds required, no tests actually running (because they hit the test cache). So the `go test` command is purely a function of disk access and CPU. For this command, Windows is consistently **15 times slower** than Linux. macOS is, pleasingly, comparable with Linux.

### Platforms affected

- [ ] Azure DevOps
- [X] GitHub Actions - Standard Runners
- [ ] GitHub Actions - Larger Runners

### Runner images affected

- [ ] Ubuntu 18.04
- [ ] Ubuntu 20.04
- [ ] Ubuntu 22.04
- [ ] macOS 11
- [ ] macOS 12
- [X] Windows Server 2019
- [X] Windows Server 2022

### Image version and build link

For `windows-2019`:

```
Current runner version: '2.303.0'
Operating System
  Microsoft Windows Server [2](https://github.com/myitcvscratch/slow-windows-actions/actions/runs/4477569194/jobs/7869257887#step:1:2)019
  10.0.1776[3](https://github.com/myitcvscratch/slow-windows-actions/actions/runs/4477569194/jobs/7869257887#step:1:3)
  Datacenter
Runner Image
  Image: windows-2019
  Version: 2023031[4](https://github.com/myitcvscratch/slow-windows-actions/actions/runs/4477569194/jobs/7869257887#step:1:4).1
  Included Software: https://github.com/actions/runner-images/blob/win19/20230314.1/images/win/Windows2019-Readme.md
  Image Release: https://github.com/actions/runner-images/releases/tag/win19%2F20230314.1
Runner Image Provisioner
  2.0.12[7](https://github.com/myitcvscratch/slow-windows-actions/actions/runs/4477569194/jobs/7869257887#step:1:8).1
```

For `windows-2022`:

```
Current runner version: '2.303.0'
Operating System
  Microsoft Windows Server [2](https://github.com/myitcvscratch/slow-windows-actions/actions/runs/4477569194/jobs/7869257680#step:1:2)022
  10.0.20[3](https://github.com/myitcvscratch/slow-windows-actions/actions/runs/4477569194/jobs/7869257680#step:1:3)[4](https://github.com/myitcvscratch/slow-windows-actions/actions/runs/4477569194/jobs/7869257680#step:1:4)8
  Datacenter
Runner Image
  Image: windows-2022
  Version: 20230314.1
  Included Software: https://github.com/actions/runner-images/blob/win22/20230314.1/images/win/Windows2022-Readme.md
  Image Release: https://github.com/actions/runner-images/releases/tag/win22%2F20230314.1
Runner Image Provisioner
  2.0.12[7](https://github.com/myitcvscratch/slow-windows-actions/actions/runs/4477569194/jobs/7869257680#step:1:8).1
````

### Is it regression?

Unclear

### Expected behavior

Windows runners to be comparable in terms of speed to Linux and macOS for `actions/checkout` and `go test` steps.

### Actual behavior

Windows runners consistently taking **15 times** as long as Linux and macOS for CPU and disk-intensive commands.

### Repro steps

The CUE repo itself is quite involved. So as a proxy for something that is relatively CPU and disk intensive we have created a slimmed down repo using `actions/checkout`.

https://github.com/myitcvscratch/slow-windows-actions

See the most recent run for results:

https://github.com/myitcvscratch/slow-windows-actions/actions/runs/4477601338

Looking at averages of this setup across a number of runs we see similar figures for `actions/checkout` to those seen in our CUE setup:

| Runner | `actions/checkout` | 
| - | - |
| `ubuntu-20.04` | 3 secs | 
| `ubuntu-22.04` | 3 secs | 
| `macos-11` | 5 secs | 
| `macos-12` | 5 secs | 
| `windows-2019` | 20 secs | 
| `windows-2022` | 20 secs |  

So whilst this doesn't include the `go test` step (because getting a warm cache is a tricky step to reproduce) the use of `actions/checkout` is a sufficiently good proxy to show the problem.

**Comment by ilia-shipitsin** [comment]
I suspect it might be Windows Defender in action.
can you try the following (in order to check whether Windows Defender disabling can help) ?

```
    - run: Set-MpPreference -DisableRealtimeMonitoring $true
      shell: powershell
```

**Comment by mvdan** [comment]
Here are many other people similarly seeing that `actions/checkout` is very slow on Windows: https://github.com/actions/checkout/issues/1150

**Comment by myitcv** [comment]
@ilia-shipitsin 

> I suspect it might be Windows Defender in action.
> can you try the following (in order to check whether Windows Defender disabling can help) ?

See commit https://github.com/myitcvscratch/slow-windows-actions/commit/8b71de30da15b600ca0d320f9415b7092e16ea6d which resulted in run https://github.com/myitcvscratch/slow-windows-actions/actions/runs/4480846656/jobs/7876717612. It has basically no effect. 

I also tried making this change in the CUE project in https://review.gerrithub.io/c/cue-lang/cue/+/551317. That resulted in https://github.com/cue-lang/cue-trybot/actions/runs/4476128503 which again showed no effect. 

So assuming my testing is valid, turning off Windows Defender does not appear to have any effect in our case.

**Comment by myitcv** [comment]
> Here are many other people similarly seeing that `actions/checkout` is very slow on Windows: [actions/checkout#1150](https://github.com/actions/checkout/issues/1150)

Thanks, @mvdan. 

Just to emphasise however that the use of `actions/checkout` in my example above is a proxy for the much bigger problem we are seeing in the `go test` step in the CUE project. It's very likely not the best proxy, but hopefully good enough. 

The reason I flag this is that ultimately we will consider _this_ issue "fixed" when the `go test` step is "fast" and not just an improvement in the `actions/checkout` step.

**Comment by ilia-shipitsin** [comment]
I've also tried to disable Windows Defender, no significant difference so far

https://github.com/ilia-shipitsin/slow-windows-actions/actions/runs/4483502669/jobs/7882868552

**Comment by ilia-shipitsin** [comment]
@myitcv , as for `go test` task performance degradation, do you see appropriate performance on standalone (inot related to GH actions) Win 2019/2022 server ?

**Comment by myitcv** [comment]
> @myitcv , as for `go test` task performance degradation, do you see appropriate performance on standalone (not related to GH actions) Win 2019/2022 server ? 
> 
> I mean, is degradation `runner specific` or `platform specific` 

We only have the numbers from GitHub actions workflow runs.

**Comment by ilia-shipitsin** [comment]
we need to narrow it, whether standalone Windows server behaves the same slow or not

**Comment by mikhailkoliada** [comment]
Duplicate of https://github.com/actions/runner-images/issues/5166

**Comment by myitcv** [comment]
@mikhailkoliada - isn't #5166 demonstrating a slowdown between 2019 and 2022? 

The numbers we are seeing show consistent slowness on both 2019 and 2022. 

Therefore I'm not clear this is a duplicate.

**Comment by ilia-shipitsin** [comment]
side note, I did some investigation on "checkout slowness", it looks like there's some delay between git and automation task. git itself takes 2-3 seconds (I put commands into cmd and wrapped with `Measure-Command { ... }`

![image](https://user-images.githubusercontent.com/125650415/226858351-7ec569c7-8d78-46cb-b6ce-72582cc39d9f.png)

I can beleive that agent communication could add 15-20 sec, but it does not look like a root cause for `go test` slowness.

**Comment by nebuk89** [comment]
@mvdan @myitcv we think this may be the same as the checkout issue. Given that 'hunch' we will tackle that tracking our progress here: https://github.com/actions/checkout/issues/1186

We will keep this issue open until we can validate if it is a dupe and go from there (or start working on a new root cause for this after I guess 😱 let's hope not!)

**Comment by myitcv** [comment]
@nebuk89 - thanks for looking into this and the detailed update. Much appreciated.

**Comment by mojca** [comment]
I would like to add that the situation seems to be even worse with larger runners. We have to use 8-core Windows runners (the regular one runs out of heap space) and they are only twice faster compared to the 2-core machines.

On Linux the 8-core vs 2-core performance is as expected (though 8-core machines are still slower in some aspects; actions like checkout take 4-times longer etc.) and the steps that can be parallelised (the build step, for example) in fact take cca. 4 times less time. On windows that ratio is around 2.

**Comment by hollowaykeanho** [comment]
providing some update here as CI tool developer:

Windows OS tends to prolong a **~5 mins** done on MacOS and Ubuntu to **1HRS+** with unknown signature (Ref: https://github.com/corygalyna/AutomataCI/actions/runs/7083465554/job/19275876384).

This is observed since November 2023 and now it is getting unusable. Some CI tasks are serial execution (CI - Materialize) and all the commands are running via local PowerShell script. 

The delay is usually at the start of a job or at the end of a job. Can't debug further.

**Comment by 1kastner** [comment]
Same issues here -https://github.com/1kastner/conflowgen/actions/runs/7202425735/job/19620518126 and https://github.com/1kastner/conflowgen/actions/runs/7202425735/job/19620518686.

It takes ages for the action to start. Before the test suite finishes (on my local laptop a matter of few minutes), it is aborted because I assume my time budget is consumed.

**Comment by cmpnnt** [comment]
This is also the case when building NuGet packages (the .NET/C# ecosystem). The process takes 57 seconds for my small project on `windows-latest` and 13 seconds on `ubuntu-latest`.

**Comment by 1kastner** [comment]
I switched from `windows-2022` to `windows-2019` and the speed increased tremendously.

**Comment by astrojuanlu** [comment]
> I switched from `windows-2022` to `windows-2019` and the speed increased tremendously.

Any chance we can get Windows XP runners?

**Comment by myitcv** [comment]
> I switched from `windows-2022` to `windows-2019` and the speed increased tremendously.

For completeness, this had no impact in our situation.

**Comment by zanieb** [comment]
Switching to `windows-2019` and disabling Windows Defender did not improve performance in my use-case.

**Comment by Somfic** [comment]
Also having issues with Windows runners, they are extremely slow compared to their MacOS and Ubuntu counterparts:

![image](https://github.com/actions/runner-images/assets/29230467/8188bc0d-80cc-4641-8a60-dc7f85b4414f)

(Can't link as the actions are in a private repo)

**Comment by harshavardhana** [comment]
Our tests are now running at 50+mins compared to Linux < 10mins, moving to 2019 didn't fix anything. We are moving away from GitHub managed runners for these, moving to in-house self-hosted runners.

**Comment by zanieb** [comment]
For what it's worth, over at Astral we switched to using a Dev Drive with ReFS to great benefit https://github.com/astral-sh/uv/pull/3522

**Comment by texadactyl** [comment]
End of July 2024 - any progress on this oldish Windows anomaly? 

My partner and I test our work on Ubuntu, MacOS, and Windows (significantly slower) to make sure we can run on all 3 environments. We are lucky that our runs are short at the moment (< 8 minutes) but expect project growth and therefore longer run times in the future.

**Comment by JacopoMadaluni** [comment]
Our benchmarks are also incredibly slow on Windows.
Our MacOS build takes about 4m (checkout, setup-node, build)
The windows counterpart takes 20+ minutes.

The build in our case is a Rust package, windows spends 10x more time to download and install crates compared to MacOS and Linux.

**Comment by stephen-carter-at-sf** [comment]
Yeah this is aweful. I also always get failures from network timeouts when on windows only. 
```
yarn install v1.22.22
[1/[5](https://github.com/forcedotcom/sfdx-scanner/actions/runs/10745601373/job/29805079241?pr=1620#step:4:6)] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
error Error: https://registry.yarnpkg.com/rxjs/-/rxjs-6.6.7.tgz: ESOCKETTIMEDOUT
...
```
It only ever happens on windows runners... never on macos or linux runners.
:-(

**Comment by texadactyl** [comment]
Not a Windoze lover but our project workfile runs on 3 OSes; Windows is the slowewwwwwwwwwwweest leg in our jobs.

cc: @platypusguy

**Comment by torokati44** [comment]
This is getting ridiculous, to the point where "optimizations" actually turn out to be "pessimizations", because of this.
See: https://github.com/ruffle-rs/ruffle/pull/17529#issuecomment-2357369221

Together with https://github.com/orgs/community/discussions/42335, not to mention https://github.com/actions/runner/issues/1182 (to potentially reduce copy-pasting somewhat, as a workaround for the workaround for the slowness...), it really makes Windows+GHA difficult to like...

**Comment by marvin-j97** [comment]
![image](https://github.com/user-attachments/assets/ecc392e7-637b-46e9-b0c1-4558637df5eb)
:smiling_face_with_tear:

**Comment by QuantumQuin** [comment]
I recently had to use a Windows runner and came here to say the experience was awful due to how slow it is. On the same size runner, it can take more than **8x** the time vs using a Mac or Linux runner. I had to use Windows to simulate a specific environment.

**Comment by xuhdev** [comment]
An interesting side note: [GitLab seems to suffer from the same issue](https://forum.gitlab.com/t/gitlab-ci-windows-shared-saas-runner-is-very-slow/92409). Maybe this is ultimately a Windows thing that is out of GitHub's control (but within GitHub's parent control :wink: ).

**Comment by subir0071** [comment]
Closing this issue as it is beyond the scope of runner images.

**Comment by ofek** [comment]
I don't think that is an appropriate response. I would expect a summary from an employee talking about precisely why there is a large discrepancy as a prerequisite for closing this topic.

**Comment by texadactyl** [comment]
Sorry if I seem dense but please explain: "Closing this issue as it is beyond the scope of runner images."

Not a Windows user at home but I know that Windows servers are potentially as fast as MacOS or Linux on the comparable hardware. So, why are Windows servers at github.com running dramatically slower than MacOS or Ubuntu? 
Is this a tuning issue? Hardware disparity or quality issue?

This issue was opened March 21, 2023. Can I assume that Microsoft would support a resolution as you are their customer?

Responses that would be helpful:
* This performance disparity will be fixed in a future project during 2025.
* We're not certain why but we are getting assistance from Microsoft. Hope to have a solution soon.

cc: @platypusguy

**Comment by texadactyl** [comment]
FWIW, the following is what Gemini responded when I asked about the disparity between windows-latest performance and that of ubuntu-latest and macos-latest.

Windows runners on GitHub Actions often exhibit significantly slower performance compared to Linux and macOS runners. There are several contributing factors to this:   

1. Operating System Overhead:

Windows Architecture: Windows, by its nature, has a more complex architecture and a larger overhead than Linux or macOS. This results in slower startup times for processes and heavier resource consumption.

Background Processes: Windows often runs more background processes and services, consuming resources that could be used for your workflow.

2. File System Performance:

NTFS vs. ext4/APFS: The NTFS file system used by Windows generally has slower performance compared to ext4 (Linux) or APFS (macOS), especially for operations involving a large number of small files, which are common in build processes.

File I/O: Windows file I/O operations can be slower, impacting tasks like package installation and compilation.

3. Software and Tooling:

Tooling Differences: Many development tools and build systems are optimized for Linux environments. Some tools might have less efficient implementations or compatibility layers on Windows.

Package Management: Package management on Windows can be slower compared to Linux's apt or yum.

Antivirus Software: Windows often has real-time antivirus software running, which can significantly impact file I/O and process execution.   

4. Virtualization and Hardware:

Virtualization Overhead: GitHub Actions runners are virtual machines. The virtualization layer on Windows might introduce more overhead than on Linux or macOS.   

Hardware Specifications: While GitHub Actions provides virtual machines with consistent specifications, there might be subtle differences in the underlying hardware or configuration that affect Windows performance.

5. Caching and Optimization:

Caching Differences: Caching mechanisms and strategies might be less effective on Windows.

Optimization: Windows workflows might require more manual optimization to achieve comparable performance to Linux or macOS.

6. Software Installation:

Software Installation Times: Installing software on Windows can often take longer than on Linux or macOS due to the nature of Windows installers and dependencies.

Mitigation Strategies:

- Use Caching: Leverage GitHub Actions caching to store dependencies and build artifacts.   
- Optimize Workflows: Streamline your workflow steps and minimize unnecessary operations.
- Use Specific Tools: Use tools and build systems that are well-optimized for Windows.
- Consider Self-Hosted Runners: If performance is critical, consider using self-hosted runners with optimized hardware and configurations.
- Minimize File I/O: Reduce the number of file I/O operations in your workflow.
- Disable Antivirus (If Possible): If your workflow allows, disable real-time antivirus scanning during the build.
- Use WSL (Windows Subsystem for Linux): For some tasks, using WSL can provide a Linux-like environment on Windows, potentially improving performance. However, this adds complexity.   
- Use Docker Containers: Using docker containers can help to create a consistent environment across all platforms.   
- By understanding these factors and applying appropriate mitigation strategies, you can improve the performance of your Windows workflows on GitHub Actions.

**Comment by mvdan** [comment]
No AI slop, thank you. Please delete that.

Does anyone know of a third party github actions runners provider that includes MacOS and Windows and provides decent performance, especially on Windows? I can't believe I'm saying this, but I don't think I can trust Microsoft as such a provider.

**Comment by cmpnnt** [comment]
> No AI slop, thank you. Please delete that.
> 
> Does anyone know of a third party github actions runners provider that includes MacOS and Windows and provides decent performance, especially on Windows? I can't believe I'm saying this, but I don't think I can trust Microsoft as such a provider.

You're welcome to spin up your own and run some tests. Let us know if your results match what GitHub is providing.

**Comment by torokati44** [comment]
@subir0071 
> Closing this issue as it is beyond the scope of runner images.

Well then would you please transfer it to the appropriate place, or open a follow-up issue there, wherever it is, and link to it here? 👼  Thank you! 🫶

**Comment by penguinland** [comment]
My team is adding Windows support to our code, and we're seeing this problem, too: the Windows build is roughly 10x slower than the Linux/Mac ones. This is not because the build requires more work: even `actions/cache@v3` is 7x slower on Windows than Linux. Maybe the Windows machines have a slower hard drive or network connection? We're eager for a workaround, if one exists.

**Comment by hudsonm62** [comment]
@penguinland another option could be to use a dev drive (added in server 2025) or a ReFS based vhdx
- https://github.com/samypr100/setup-dev-drive
- https://chadgolden.com/blog/github-actions-hosted-windows-runners-slower-than-expected-ci-and-you#so-what-else-can-we-do-harder-wins

---

Still, it is quite infuriating the core issue can't be looked at or investigated more... @nebuk89 any chance of this being re-opened and/or properly investigated? Whilst workarounds exist, its much harder to implement in matrix flows.

**Comment by mvdan** [comment]
For those following along and who might find this useful - we ended up switching to https://namespace.so/, who started offering Windows runners recently. For the same price per minute as github.com, we have seen a 2-3x improvement in CPU speed, way faster caching via mounted volumes, and we haven't run into any issues with disk or filesystem speeds.

**Comment by Frulfump** [comment]
> [@penguinland](https://github.com/penguinland) another option could be to use a dev drive or a ReFS based vhdx
> 
>     * https://github.com/samypr100/setup-dev-drive
> 
>     * https://chadgolden.com/blog/github-actions-hosted-windows-runners-slower-than-expected-ci-and-you#so-what-else-can-we-do-harder-wins
> 
> 
> Still, it is quite infuriating the core issue can't be looked at or investigated more... [@nebuk89](https://github.com/nebuk89) any chance of this being re-opened and/or properly investigated? Whilst workarounds exist, its much harder to implement in matrix flows.

Yeah very annoying, a bit strange that the agents don't use Dev Drive since Windows I/O is so bad compared to Linux and macOS.

**Comment by MatthewSteeples** [comment]
@Frulfump https://github.com/actions/runner-images/issues/8698
